### PR TITLE
RH2117972 - Extend the support for NSS DBs (PKCS11) in FIPS mode

### PR DESCRIPTION
**Pull request by martinuy**
_[Search this PR in Red Hat Jira](https://issues.redhat.com/issues/?jql=issueFunction%20in%20linkedIssuesOfRemote(url,"https://github.com/rh-openjdk/jdk/pull/17"))_

Initial proposal for this enhancement. Find the full description of the changes here: https://bugzilla.redhat.com/show_bug.cgi?id=2117972

**Important note:** this enhancement requires the following additional changes to be shipped in RPM builds:

 * nss.fips.cfg (replace the attributes below)
   * nssSecmodDirectory = ${fips.nssdb.path}
   * nssDbMode = readWrite

### CODE REVIEW

**Comment by martinuy** [comment]
The test jdk/java/util/logging/SimpleFormatterFormat.java (linux-x64-fips jdk tier-1 part-2) failed during the pre-submit tests. After analyzing the test artifacts, it seems that something was thrown to stderr and the test, who was recording stderr to match against its expected strings, got an unexpected line. The unexpected line has the following content: 'properties: reading security properties file: /home/runner/jdk-linux-x64/jdk-17.0.3-internal+0_linux-x64_bin/jdk-17.0.3/conf/security/java.security'. This line is recorded by the method 'initialize' in java.security.Security, and it's not specific to FIPS: it's found in OpenJDK upstream too. I don't see how this test could be related to the changes proposed in the context of this PR, or to FIPS in general. If I had to guess, it could be that the class java.security.Security was being initialized in parallel to the test and its debugging output got in the way of the test's stderr recording. I executed it locally multiple times in a FIPS environment and succeeded. I'll resubmit the tests job but I don't currently see a reason to put this PR-review on hold.

**Comment by martinuy** [comment]
Update: the jdk/java/util/logging/SimpleFormatterFormat.java test failed again in the 2nd run but I noticed that it fails in other branches too, such as this one [1]. In conclusion, this failure is not a regression caused by this PR.

--
[1] - https://github.com/rh-openjdk/jdk/runs/7810906645?check_suite_focus=true

**Comment by franferrax** [comment]
> Update: the jdk/java/util/logging/SimpleFormatterFormat.java test failed again in the 2nd run but I noticed that it fails in other branches too, such as this one [1]. In conclusion, this failure is not a regression caused by this PR.
> 
> --
> [1] - https://github.com/rh-openjdk/jdk/runs/7810906645?check_suite_focus=true

@martinuy: this appeared in #11, as explained in [this comment](https://github.com/rh-openjdk/jdk/pull/11#issuecomment-1165744413). It's caused by the `-Djava.security.debug=properties` usage in the *Linux x64 + FIPS* execution.

**Location in submit.yml:** `jobs` / `linux_x64_test_fips` *(Linux x64 + FIPS)* / `steps` / `#12` *(Run tests)*
https://github.com/rh-openjdk/jdk/blob/03b584eb4a2981eea0bd0660bcdeb7490c5bf6e0/.github/workflows/submit.yml#L498

**Review by franferrax — COMMENTED** [review]
@martinuy: I will post several comments in line with the code, but please note that none of them are blocking ones. Feel free to close every conversation as you prefer:
 * Without any answer: I'll just assume you acknowledged a minor comment
 * Exposing your point/opinion: since no comment is blocking, this is fine to me
 * Changing the code, if you agree with me and find the comment worth taking action

I think there'll be some suggestions for the `NssdbPin.java` test that may slightly simplify the code or improve its readability, so I expect some of them to be considered.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (franferrax) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java:55]: I was wondering whether it makes sense to add `assert callbacks.length > 0;`, but to me, it also seems safe to assume `callbacks` will hold at least an element.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (franferrax) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java:89]: Are we ok with `InvalidPathException` (unchecked) being uncaught in case the path is invalid? Seems reasonable to me, but just wanted to ensure you were aware of this.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (franferrax) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java]: Typo: right phrasing: 'in the same way **as** NSS modutil'

File: src/java.base/share/conf/security/java.security
Comment (franferrax) [inline src/java.base/share/conf/security/java.security]: It seems that this paragraph refers to the `pin:<value>` usage. I see that both environment variables [\[1\]](#2c8a30d853997ec996c94347f35eea4d) (`env:<value>` usage) and files [\[2\]](#8f03fa5bae9b2209467d14eeaa23742d) (`file:<value>` usage) are decoded from UTF-8, but shouldn't we clarify this?

----
**<a id="2c8a30d853997ec996c94347f35eea4d"></a>\[1\]** Python command showing how environment variables with different encodings are displayed by `bash`, suggesting the right encoding is UTF-8:
```python
python3 -BISc "
import os
import shutil

bash_cmd = 'echo Unicode BULLET with different encodings:'
encodings = 'cp1252', 'big5', 'utf-7', 'utf-8'
for idx, encoding in enumerate(encodings):
    os.putenv(f'VAR_{idx}', '\N{BULLET}'.encode(encoding))
    bash_cmd += f' && echo - {encoding}: \N{DOLLAR SIGN}VAR_{idx}'

bash = shutil.which('bash')
os.execv(bash, [bash, '-c', bash_cmd])
"
```
Output:
```
Unicode BULLET with different encodings:
- cp1252: �
- big5: �E
- utf-7: +ICI-
- utf-8: •
```

----
**<a id="8f03fa5bae9b2209467d14eeaa23742d"></a>\[2\]** Snippet where the file is read
https://github.com/rh-openjdk/jdk/blob/316e9e35330671bfff009f3c1b3d35c449108af0/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java#L139-L143

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: Typo: **performed**

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: Isn't this just `"1234567890abcdef1234567890ABCDEF\uA4F7"`? The following snippet gives me `$1 ==> true` in a `jshell`:
```java
"1234567890abcdef1234567890ABCDEF\uA4F7".equals(
    "1234567890abcdef1234567890ABCDEF" + new String(new char[] {'\uA4F7'})
)
```

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java:75]: This seems to have been forgotten as `true`, otherwise, I don't see a reason for `DEBUG` to exist

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java:94]: This should never happen, but if the test is executed with a single argument (or two), that would cause an `ArrayIndexOutOfBoundsException` when accessing `args[1]` (or `args[2]`), that's why I'm slightly inclined towards an `if (args.length == 3) {...}` check

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: We don't need to call `loginType.name()` since the implicit `loginType.toString()` in `"Login type: " + loginType` would shield the same result

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java:237]: Maybe an `assert ctx.nssdbPinFile != null;`, and/or a comment saying that the file is written by `createNSSDB(ctx)` can help the reader here

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java:246]: We can do this because `explicitLoginLauncher()` is called after using the `ctx.loginType` initialization by `TestContext.<init>()` (as `LoginType.IMPLICIT`). If we move `explicitLoginLauncher()` before the `pinLauncher()`/`envLauncher()`/`fileLauncher()` loop, we would easily ruin the `ctx.loginType` value. I'm not sure about how to do this more symmetric or have all the `ctx.loginType` and `ctx.propType` writes closer.

Maybe it is clearer to set `ctx.loginType` and `ctx.propType` for all the cases in `mainLauncher()`, removing these two lines and the `ctx.loginType` initialization in `TestContext.<init>()`.

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: `ctx.nssdbPath` would be preciser than `ctx.workspace` here

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java:333]: From the second usage of `getModutilPB()`, `opts = Arrays.asList("-changepw", NSSDB_TOKEN_NAME)` and `args = Arrays.asList("-newpwfile", ctx.nssdbPinFile.toString())`. I don't see how the first are 'options' while the second are 'arguments', in my opinion, it makes no sense to have `opts` and `args` separated if they don't have any distinctive semantic.

I would rather unify them as `args`, removing the need to pass `null` in the first `getModutilPB()` call, and I would also employ variadic parameters to remove the need of using `Arrays.asList()` (which also seems to match the decision made for [the `ProcessBuilder` constructor we are using](/rh-openjdk/jdk/blob/bb46af07cb95d090d53159a2e71be9b72fb581f8/src/java.base/share/classes/java/lang/ProcessBuilder.java#L217-L232)).

**Comment by gnu-andrew** [comment]
Please don't merge this until the other pending FIPS changes are out of the way or we'll have to maintain two branches.

We should also look at getting `nss.fips.cfg` into the repository so changes like this are self-contained.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java:55]: Yes, I don't deem it necessary as the only caller (SunPKCS11::login) always passes a handler. A login attempt without that would not make sense and would immediately expose the error. I don't see harm in adding the assertion either, but the assumption looks safe and obvious to me.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java:89]: Should be a rare event but in case it happens we cannot proceed so throwing an unchecked exception sounds reasonable to me. This is what we do already with ProviderException (which is another type of RuntimeException, like InvalidPathException). I don't see much value in wrapping the InvalidPathException in a ProviderException; the user might be familiar with InvalidPathException if a path is wrong. The stack trace will point exactly to the error.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java]: Good, will be fixed.

File: src/java.base/share/conf/security/java.security
Comment (martinuy) [reply src/java.base/share/conf/security/java.security]: The comment applies to the security property independently if it's used to pass a PIN value, a file path or an environment variable. The underlying problem is that _java.security_ does not accept encodings such as UTF-8. I'm not sure of the reasons but probably historical. To workaround this we can escape _unicode_ characters with the _\u_ syntax. We can isolate this problem with this code:

```java
public final class Main {
    private static final String CHILD_ID = "CHILD_ID";
    private static final String SEC_PROP = "SEC_PROP";
    private static java.util.List<String> SEC_PROP_VALS =
            java.util.Arrays.asList(new String[] {"ú", "\\u00FA"});
    public static void main(String[] args) throws Throwable {
        if (args.length > 1 && args[0].equals(CHILD_ID)) {
            String secVal = java.security.Security.getProperty(SEC_PROP);
            System.out.println("SEC_PROP (passed): " + args[1] + " - SEC_PROP (received): "
                    + secVal);
        } else {
            for (String secPropVal : SEC_PROP_VALS) {
                Proc p = Proc.create(Main.class.getName()).args(CHILD_ID, secPropVal);
                p.secprop(SEC_PROP, secPropVal);
                p.inheritIO();
                p.start().waitFor(0);
            }
        }
    }
}
```

This is the output:

`SEC_PROP (passed): ú - SEC_PROP (received): Ãº`
`SEC_PROP (passed): \u00FA - SEC_PROP (received): ú`

I'l leave this comment unresolved until we make sure that we are on the same page.

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: Good, will be fixed

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: Good, will be improved

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:237]: I'll add a comment if it makes things easier for the reader.

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: That's right. Will be fixed

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:246]: Yes, I went through this same dilemma between being more explicit and having fewer code lines. I'll explicitly set the `ctx.loginType` to `LoginType.IMPLICIT` before the loop, but not per-iteration. We can infer from the context that it won't be modified by the methods called inside the loop.  I'll add this block into its own scope, so it's clear where to put `explicitLoginLauncher` if relocated.

The protocol would be: use the _ctx_ but make sure that `loginType` and `propType` are set to the right values as they are mutable. It was more of a convenience to pass values internally within the class methods, I wouldn't add more complexity.

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:333]: The distinction between arguments and options matters (and is documented) at the _modutil_ level because of the order: you cannot pass an argument before an option. I'll leave this comment unresolved in case that you want to add something else to this conversation.

File: src/java.base/share/conf/security/java.security
Comment (franferrax) [reply src/java.base/share/conf/security/java.security]: You are right, I got a bit confused, we are always talking about `<value>`, not the PIN (which only happens to be `<value>` in the `pin:<value>` case).

If you agree that it would be clearer to the user, rephrase to explicitly refer to `<value>`, I'm thinking something like:
```
# When used as a system property, encode <value> in UTF-8. When used as
# a security property (i.e. in this file), encode non-Basic Latin Unicode
# characters with \uXXXX.
```

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:333]: Ah ok, so the semantic is in `modutil` itself. If I only pay attention to the _Java_ code, I still think that having `getModutilPB()` calls like the following would read similar enough to a `modutil` command in the terminal, to make them self explanatory.
```java
pb = getModutilPB(ctx, "-create");
//--> modutil -force -create -dbdir <nssdbPath>

pb = getModutilPB(ctx, "-changepw", NSSDB_TOKEN_NAME, "-newpwfile", ctx.nssdbPinFile.toString());
//--> modutil -force -changepw <NSSDB_TOKEN_NAME> -newpwfile <nssdbPinFile> -dbdir <nssdbPath>
```


Also, since the current code doesn't handle them in any particular way, we can have the following:
```java
// Right
ProcessBuilder pb = getModutilPB(ctx, Arrays.asList("-create"), null);

// Semantically wrong, but works exactly the same
ProcessBuilder pb = getModutilPB(ctx, null, Arrays.asList("-create"));
```

I just want to make my point clear, but I repeat that none of this blocks the review approval though.

File: src/java.base/share/conf/security/java.security
Comment (martinuy) [reply src/java.base/share/conf/security/java.security]: Okay, I'll add something in the comment

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java]: Good, will be fixed

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:75]: Yes, will turn that off now

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:94]: Good, will be fixed

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:333]: Yes, I see your point. We can make this change if the function getModutilPB only adds options at the beginning and arguments at the end, and if the caller gives an ordered list. Will implement the change.

**Comment by martinuy** [comment]
Hi @franferrax ,

Thanks for your review. A new changeset has been pushed for you to have another look.

Martin.-

**Review by franferrax — DISMISSED** [review]
@martinuy: I've reviewed b7119b52965d287768a26b8b7cce3af7e093b8ea, everything looks good to me. Thank you for taking action even with minor comments!

**Comment by gnu-andrew** [comment]
Once the PKCS#12 and TLSv1.3 changes are settled, I'll open a PR to get `nss.fips.cfg` into the tree. This can then be updated with the necessary changes for that file too.

Please wait for that before merging this. Thanks.

**Comment by gnu-andrew** [comment]
nss.fips.cfg added in #22

Once this change is rebased on top of that with the `nss.fips.cfg` changes, we should be able to get this in.

**Comment by martinuy** [comment]
I've rebased this PR to include https://github.com/rh-openjdk/jdk/pull/22

The rebase was clean.

After the rebase, I included the required changes to `nss.fips.cfg` as part of this PR. See [here](https://github.com/rh-openjdk/jdk/pull/17/commits/f096804b619833f126a48c01da590ecd37d9cd67). The test `jdk/sun/security/pkcs11/fips/NssdbPin.java` passed.

Can I have a new review?

Thanks.-

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (gnu-andrew) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java]: So, in the event that we fail to read a PIN, the PIN will be an empty char array. That's different from the `null` we have in the non-FIPS case. Is that deliberate?

File: src/java.base/share/conf/security/java.security
Comment (gnu-andrew) [inline src/java.base/share/conf/security/java.security:333]: The prefixes are listed in lower-case here, but the checks are case-sensitive and all caps.
We should either fix the documentation or make the check case-insensitive.

**Review by gnu-andrew — COMMENTED** [review]
Just a couple of things from me which should be quick to resolve. Thanks for the quick rebase.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java]: The non-FIPS `provider.login(null, null)` statement here [1] cannot be directly compared to the return statement here [2] because the former implies not calling PKCS#11 `C_Login` in the token [3] (which is fine if the NSS DB does not require a login), while the latter implies calling it with a PIN of len == 0 [4] (which must be done if the NSS DB requires a login and a PIN has not been initialized in the DB). The effect of returning null or empty-array here [2] is whether we call `C_Login` with a NULL `pPin` parameter here [5] or not, but in both cases `ulPinLen` is 0 and that's what matters for our default FIPS scenario. Given that the effect is the same but we can save some performance penalty by not creating an empty array, I will make the change.

Let me take this opportunity to give some context and explain why we want to always do a login in FIPS mode (which is calling `provider::login` with a non-null 2nd parameter) [6].

If an NSS DB is not configured to require a login, the NSS Software Token cannot operate in FIPS-140 level 2 [7]. Please keep in mind that OpenJDK and the NSS Software Token must operate in FIPS-140 level 2 when RHEL is in FIPS mode. Thus, not making a login is not right.

There are 2 ways in which an NSS DB can require a login: not having a PIN set (think of this as an unitialized PIN) [8] or having a PIN set to something different than empty string [9]. Thus, having an empty string PIN set in the NSS DB does not require a login. When the NSS DB requires a login, the internal `CKF_LOGIN_REQUIRED` flag is set.

RHEL's deafult NSS DB (`/etc/pki/nssdb`) does not have a PIN set. Thus, the `CKF_LOGIN_REQUIRED` flag is set and a login is required. To login in this DB we must use an empty string PIN (from the user's new API point of view) or null PIN (from the `C_Login` point of view). In other words, `/etc/pki/nssdb` requires an empty string login, which is calling the PKCS#11 `C_Login` API with `usPinLen == 0` [10], and the underlying `NSC_Login` with `ulPinLen == 0` [5].

In summary, the change I made [here](https://github.com/rh-openjdk/jdk/pull/17/commits/01bf0e83b42c5bba2c5bdcdf3cae7ce3a7b21d36) did not cause any visible effect but I adjusted the test case assertion a bit to handle the new null value. I also debugged `C_Login` in the NSS Software Token side while opening `/etc/pki/nssdb` to make sure that we are exercising the case with the changes.

--
[1] - https://github.com/rh-openjdk/jdk/pull/17/files#diff-fb2c3293fab6ce24717b91fcf63fb9b33989bce9518910faf46e086493c753a3R247
[2] - https://github.com/rh-openjdk/jdk/pull/17/files#diff-dfa0dff6275fe003a6a2b68903fe3b3d9b0fac72d988f50a3554cd533c9efbd9R122
[3] - https://github.com/rh-openjdk/jdk/blob/f096804b619833f126a48c01da590ecd37d9cd67/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java#L1668
[4] - https://github.com/rh-openjdk/jdk/blob/f096804b619833f126a48c01da590ecd37d9cd67/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java#L1716
[5] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/pkcs11.c#L4289
[6] - https://github.com/rh-openjdk/jdk/pull/17/files#diff-fb2c3293fab6ce24717b91fcf63fb9b33989bce9518910faf46e086493c753a3R244
[7] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/fipstokn.c#L638
[8] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/pkcs11.c#L3720
[9] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/pkcs11.c#L3724
[10] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/fipstokn.c#L876

File: src/java.base/share/conf/security/java.security
Comment (martinuy) [reply src/java.base/share/conf/security/java.security:333]: While the checks are case-sensitive, we make the obtained prefix uppercase  [here](https://github.com/rh-openjdk/jdk/pull/17/files#diff-dfa0dff6275fe003a6a2b68903fe3b3d9b0fac72d988f50a3554cd533c9efbd9R75), before any comparison. Regarding documenting the prefix as lower or uppercase, I prefer lowercase for better alignment with NSS tools. The current `NssdbPin.java` test should be exercising this because it's passing lowercase, as documented for users.

**Comment by martinuy** [comment]
@gnu-andrew Thanks for your review! Please have a look at the inlined comments and the new change.

@franferrax I'd be grateful if you can also have a look, as your previous approval is now invalid because of the changes.

File: src/java.base/share/conf/security/java.security
Comment (martinuy) [reply src/java.base/share/conf/security/java.security:333]: I made the change to lowercase anyways, for better alignment between the code and the doc. Thanks.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java
Comment (gnu-andrew) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java]: Thanks for the detailed explanation. I remember us adding the code for the empty login, so I recall why that was required.

I agree returning `null` is a little more efficient and also clearer, so thanks for making that change. The new `if` statement in the test case could be a little clearer with more bracketing, but that's not a blocker.

File: src/java.base/share/conf/security/java.security
Comment (gnu-andrew) [reply src/java.base/share/conf/security/java.security:333]: Thanks, I missed the case conversion, which is more efficient than the `equalsIgnoreCase` I was thinking of. Thanks for syncing the docs with the code.

**Review by gnu-andrew — APPROVED** [review]
Looks good with the changes made, though the new `if` block could be a little clearer with more bracketing (I'll leave that to your personal choice)

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (franferrax) [inline test/jdk/sun/security/pkcs11/fips/NssdbPin.java:184]: Maybe I'm missing something, but couldn't we just have declared `String pin = "";` instead of `String pin = null;` here?

Wouldn't it have allowed us to keep the `!pin.equals(expectedPin)` previous (simpler) condition in the following `if` block? NOTE: with 'previous' I mean the left side of f9afab177d4b09e1e23a79ebe311d978101c416a...69cabe49b1802876a4b277ea2f99742245436652.

**Review by franferrax — APPROVED** [review]
> @franferrax I'd be grateful if you can also have a look, as your previous approval is now invalid because of the changes.

@martinuy: I reviewed the new changes since my last review (f9afab177d4b09e1e23a79ebe311d978101c416a...69cabe49b1802876a4b277ea2f99742245436652, taking into account that the rebase was clean). Everything seems fine to me, but I added a minor comment for the test, which at the same time would address [Andrew's comment](https://github.com/rh-openjdk/jdk/pull/17#pullrequestreview-1188466624).

The review is still approved since, in my view, suggestions for tests aren't blockers unless the test is incorrect or the suggested code has really noticeable advantages in performance, readability, etc.

File: test/jdk/sun/security/pkcs11/fips/NssdbPin.java
Comment (martinuy) [reply test/jdk/sun/security/pkcs11/fips/NssdbPin.java:184]: The `pin` local variable initialized to `null` allows us to distinguish whether the `getFipsNssdbPin` returned empty string or null, which is the change that we just made and the behavior that we expect.

**Comment by martinuy** [comment]
> Looks good with the changes made, though the new `if` block could be a little clearer with more bracketing (I'll leave that to your personal choice)

Thanks for your review. Yes, I sometimes lean towards more bracketing but in this case I thought that relying on the precedence between the `or` and the `and` was enough.

**Comment by gnu-andrew** [comment]
> > Looks good with the changes made, though the new `if` block could be a little clearer with more bracketing (I'll leave that to your personal choice)
> 
> Thanks for your review. Yes, I sometimes lean towards more bracketing but in this case I thought that relying on the precedence between the `or` and the `and` was enough.

It is for the computer. It just took longer for my brain to work out what was going on :)

**Pull request merged**
